### PR TITLE
Ignore dynamic vs object, etc in pattern-matching machinery.

### LINKED ISSUE #31269: Should ignore dynamic vs object in subsumption tests.
The following test fails

``` c#
        [Fact]
        [WorkItem(16195, "https://github.com/dotnet/roslyn/issues/16195")]
        public void TestIgnoreDynamicVsObjectAndTupleElementNames_01()
        {
            var source =
@"public class Generic<T>
{
    public enum Color { Red, Blue }
}
class Program
{
    public static void Main(string[] args)
    {
    }
    public static void M2(object o)
    {
        switch (o)
        {
            case Generic<long>.Color c:
            case Generic<object>.Color.Red:
            case Generic<(int x, int y)>.Color.Blue:
            case Generic<string>.Color.Red:
            case Generic<dynamic>.Color.Red: // error: duplicate case
            case Generic<(int z, int w)>.Color.Blue: // error: duplicate case
            case Generic<(int z, long w)>.Color.Blue:
            default:
                break;
        }
    }
}
";
            CreateCompilation(source).VerifyDiagnostics(
                // (18,13): error CS8120: The switch case has already been handled by a previous case.
                //             case Generic<dynamic>.Color.Red: // error: duplicate case
                Diagnostic(ErrorCode.ERR_SwitchCaseSubsumed, "case Generic<dynamic>.Color.Red:").WithLocation(18, 13),
                // (19,13): error CS8120: The switch case has already been handled by a previous case.
                //             case Generic<(int z, int w)>.Color.Blue: // error: duplicate case
                Diagnostic(ErrorCode.ERR_SwitchCaseSubsumed, "case Generic<(int z, int w)>.Color.Blue:").WithLocation(19, 13)
                );
        }
```

### PR DESCRIPTION
**Pull request by gafter**
Fixes #31269

### CODE REVIEW

File: src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs [hunk 1]
```diff
@@ -1,15 +1,22 @@
...
+        private static IEqualityComparer<ISymbol> s_ignoreEverythingComparer = SymbolEqualityComparer.IgnoreEverything;
+
         public override bool Equals(object obj) => obj is BoundDagEvaluation other && this.Equals(other);
         public virtual bool Equals(BoundDagEvaluation other)
         {
-            return other != (object)null && this.Kind == other.Kind && this.GetOriginalInput().Equals(other.GetOriginalInput()) && this.Symbol == other.Symbol;
+            return other != (object)null &&
+                this.Kind == other.Kind &&
+                this.GetOriginalInput().Equals(other.GetOriginalInput()) &&
+                s_ignoreEverythingComparer.Equals(this.Symbol?.GetPublicSymbol(), other.Symbol?.GetPublicSymbol());
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs]: >s_ignoreEverythingComparer.Equals(this.Symbol?.GetPublicSymbol(), other.Symbol?.GetPublicSymbol() [](start = 16, length = 97)

Do we need to use `SymbolEqualityComparer` and `GetPublicSymbol()` or could this be `Symbol.Equals(this.Symbol, other.Symbol, TypeCompareKind.AllIgnoreOptions)`?

File: src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs [hunk 2]
```diff
@@ -26,9 +33,10 @@ private Symbol Symbol
                 }
             }
         }
+
         public override int GetHashCode()
         {
-            return Hash.Combine(GetOriginalInput().GetHashCode(), this.Symbol?.GetHashCode() ?? 0);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs]: >this.Symbol?.GetHashCode() ?? 0) [](start = 66, length = 32)

Is this change needed?

Comment (gafter) [reply src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs]: Yes.  We need to treat an evaluation of `X<object>.Field` as the same field as `X<dynamic>.Field`.

---
In reply to: [362567115](https://github.com/dotnet/roslyn/pull/40677#discussion_r362567115) [](ancestors = 362567115)

**Comment by gafter** [comment]
@cston I've responded to your comments.  Do you have any others?

@RikkiGibson Could you review this please?

File: src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs]: I believe we ignore differences in `object` and `dynamic` in `GetHashCode()`  (see `DynamicTypeSymbol.GetHashCode()` for instance) so the hash codes for `X<object>.Field` and `X<dynamic>.Field` should be the same. 

And, it looks like `SymbolEqualityComparer.GetHashCode(ISymbol obj)` simply calls `obj.UnderlyingSymbol.GetHashCode()` as the previous code here was doing.

---
In reply to: [362689732](https://github.com/dotnet/roslyn/pull/40677#discussion_r362689732) [](ancestors = 362689732,362567115)

File: src/Compilers/Core/Portable/Symbols/SymbolEqualityComparer.cs [hunk 3]
```diff
@@ -26,6 +26,9 @@ public sealed class SymbolEqualityComparer : IEqualityComparer<ISymbol?>
         // Internal only comparisons:
         internal readonly static SymbolEqualityComparer ConsiderEverything = new SymbolEqualityComparer(TypeCompareKind.ConsiderEverything);
 
+        // Ignore everything; for types, this reflects the identity conversion.
+        internal readonly static SymbolEqualityComparer IgnoreEverything = new SymbolEqualityComparer(TypeCompareKind.AllIgnoreOptions);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/SymbolEqualityComparer.cs]: >internal readonly static SymbolEqualityComparer IgnoreEverything = new SymbolEqualityComparer(TypeCompareKind.AllIgnoreOptions); [](start = 8, length = 128)

Compilers should not be using comparers that operate on ISymbol interfaces. CSharp compiler has its own symbol comparer that already provides this flavor for internal symbols.

File: src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs [hunk 4]
```diff
@@ -1,15 +1,25 @@
...
 
+#nullable enable
+
 namespace Microsoft.CodeAnalysis.CSharp
 {
     partial class BoundDagEvaluation
     {
-        public override bool Equals(object obj) => obj is BoundDagEvaluation other && this.Equals(other);
-        public virtual bool Equals(BoundDagEvaluation other)
+        private static IEqualityComparer<ISymbol> s_ignoreEverythingComparer = SymbolEqualityComparer.IgnoreEverything;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs]: >private static IEqualityComparer<ISymbol> s_ignoreEverythingComparer = SymbolEqualityComparer.IgnoreEverything; [](start = 8, length = 111)

We shouldn't be using this comparer in compilers, there is a ```Microsoft.CodeAnalysis.CSharp.Symbols.SymbolEqualityComparer``` if we really need to use a comparer.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs]: That class also keeps singletons for commonly used comparers

---
In reply to: [362895287](https://github.com/dotnet/roslyn/pull/40677#discussion_r362895287) [](ancestors = 362895287)

File: src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs [hunk 5]
```diff
@@ -26,9 +36,10 @@ private Symbol Symbol
                 }
             }
         }
+
         public override int GetHashCode()
         {
-            return Hash.Combine(GetOriginalInput().GetHashCode(), this.Symbol?.GetHashCode() ?? 0);
+            return Hash.Combine(GetOriginalInput().GetHashCode(), s_ignoreEverythingComparer.GetHashCode(Symbol.GetPublicSymbol()));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs]: >s_ignoreEverythingComparer.GetHashCode(Symbol.GetPublicSymbol()) [](start = 66, length = 64)

This change is not needed. All symbols are supposed to implement GetHashCode in a way that doesn't pay attention to all aspects that could ever be ignored by different flavors of TypeCompareKind. Comparer simply delegates back to the symbol.

File: src/Compilers/CSharp/Portable/BoundTree/BoundDagTemp.cs [hunk 6]
```diff
@@ -1,12 +1,15 @@
...
 
+using System.Collections.Generic;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
     partial class BoundDagTemp
     {
+        private static IEqualityComparer<ISymbol> s_ignoreEverythingComparer = SymbolEqualityComparer.IgnoreEverything;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundDagTemp.cs]: > private static IEqualityComparer<ISymbol> s_ignoreEverythingComparer = SymbolEqualityComparer.IgnoreEverything; [](start = 7, length = 112)

Same comment as for BoundDagEvaluation

File: src/Compilers/CSharp/Portable/BoundTree/BoundDagTemp.cs [hunk 7]
```diff
@@ -15,18 +18,24 @@ partial class BoundDagTemp
...
-            return other != (object)null && this.Type.Equals(other.Type, TypeCompareKind.AllIgnoreOptions) && object.Equals(this.Source, other.Source) && this.Index == other.Index;
+            return other != (object)null &&
+                this.Type.Equals(other.Type, TypeCompareKind.AllIgnoreOptions) &&
+                object.Equals(this.Source, other.Source) && this.Index == other.Index;
         }
+
         public override int GetHashCode()
         {
-            return Hash.Combine(this.Type.GetHashCode(), Hash.Combine(this.Source?.GetHashCode() ?? 0, this.Index));
+            return Hash.Combine(s_ignoreEverythingComparer.GetHashCode(this.Type.GetPublicSymbol()), Hash.Combine(this.Source?.GetHashCode() ?? 0, this.Index));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundDagTemp.cs]: >s_ignoreEverythingComparer.GetHashCode(this.Type.GetPublicSymbol()) [](start = 32, length = 67)

Same comment as for BoundDagEvaluation

File: src/Compilers/Core/Portable/Symbols/SymbolEqualityComparer.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Symbols/SymbolEqualityComparer.cs]: After looking through remaining parts of this PR, I think that this class shouldn't be modified.

---
In reply to: [362893135](https://github.com/dotnet/roslyn/pull/40677#discussion_r362893135) [](ancestors = 362893135)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

**Comment by gafter** [comment]
@cston @AlekseyTs I believe I've responded to all of your commens.  Do you have any others?

**Pull request merged**